File: kotlin-orders/kotlin-orders-dto/src/test/java/com/mvv/bank/orders/domain/TradesTest.kt
package com.mvv.bank.orders.domain

import com.mvv.bank.orders.domain.test.TestDateTimeService
import com.mvv.bank.orders.domain.test.predefined.TestPredefinedCompanies
import com.mvv.bank.orders.domain.test.predefined.TestPredefinedMarkets
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions
import org.junit.jupiter.api.Test
import java.math.BigDecimal as bd
import java.time.ZonedDateTime


internal class TradesTest {
    private val now = ZonedDateTime.now()

    @Test
    fun createFxCashTrade() {
        val trade = FxCashTrade(
            id = 123L,
            market = TestPredefinedMarkets.KYIV1.symbol,
            buySellType = BuySellType.BUY,
            buyCurrency = Currency.EUR,
            sellCurrency = Currency.UAH,
            volume = bd("1000"),
            price = amount("36.7 UAH"),
            tradedAt = now,
        )

        SoftAssertions().apply {

            assertThat(trade.id).isEqualTo(123L)
            assertThat(trade.market).isNotNull.isEqualTo(TestPredefinedMarkets.KYIV1.symbol)
            assertThat(trade.buySellType).isEqualTo(BuySellType.BUY)
            assertThat(trade.buyCurrency).isEqualTo(Currency.EUR)
            assertThat(trade.sellCurrency).isEqualTo(Currency.UAH)

            assertThat((trade as Trade<Currency>).product).isEqualTo(Currency.EUR)
            assertThat(trade.priceCurrency).isEqualTo(Currency.UAH)

            assertThat(trade.volume).isEqualTo(bd(1000))
            assertThat(trade.price).isEqualTo(amount("36.7 UAH"))
            assertThat(trade.tradedAt).isEqualTo(now)

        }.assertAll()
    }

    @Test
    fun createStockTrade() {
        val trade = StockTrade(
            id = 124L,
            market = TestPredefinedMarkets.NASDAQ.symbol,
            company = TestPredefinedCompanies.APPLE.symbol,
            buySellType = BuySellType.BUY,
            volume = bd("1000"),
            price = amount("95.55 USD"),
            tradedAt = now,
        )

        SoftAssertions().apply {

            assertThat(trade.id).isEqualTo(124L)
            assertThat(trade.market).isNotNull.isEqualTo(TestPredefinedMarkets.NASDAQ.symbol)
            assertThat(trade.buySellType).isEqualTo(BuySellType.BUY)
            assertThat(trade.product).isEqualTo(TestPredefinedCompanies.APPLE.symbol)
            assertThat(trade.company).isEqualTo(TestPredefinedCompanies.APPLE.symbol)
            assertThat(trade.volume).isEqualTo(bd(1000))
            assertThat(trade.price).isEqualTo(amount("95.55 USD"))
            assertThat(trade.tradedAt).isEqualTo(now)

        }.assertAll()
    }

    @Test
    fun testTestDateTimeService() {
        val dateTimeService = TestDateTimeService(nowSupplier = { now.toInstant() })
        assertThat(dateTimeService.now()).isEqualTo(now.toInstant())
    }
}

File: scala-orders/src/test/scala/com/mvv/bank/orders/domain/TradesTest.scala
package com.mvv.bank.orders.domain

import scala.language.unsafeNulls
//
import scala.math.BigDecimal as bd
//
import java.time.ZonedDateTime
//
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.SoftAssertions
import org.junit.jupiter.api.Test
//
import com.mvv.bank.orders.domain.test.predefined.{ TestPredefinedCompanies, TestPredefinedMarkets, TestDateTimeService }
import com.mvv.test.SoftAssertions.runTests


class TradesTest {
    private val now = ZonedDateTime.now()

    @Test
    def createFxCashTrade():Unit = {
        val trade = FxCashTrade(
            id = 123L,
            market = TestPredefinedMarkets.KYIV1.symbol,
            buySellType = BuySellType.BUY,
            buyCurrency = Currency.EUR,
            sellCurrency = Currency.UAH,
            volume = bd("1000"),
            price = Amount.of(bd("36.7"), Currency.UAH),
            tradedAt = now,
        )

        SoftAssertions().runTests { a =>

            a.assertThat(trade.id).isEqualTo(123L)
            a.assertThat(trade.market).isNotNull.isEqualTo(TestPredefinedMarkets.KYIV1.symbol)
            a.assertThat(trade.buySellType).isEqualTo(BuySellType.BUY)
            a.assertThat(trade.buyCurrency).isEqualTo(Currency.EUR)
            a.assertThat(trade.sellCurrency).isEqualTo(Currency.UAH)

            a.assertThat(trade.asInstanceOf[Trade[Currency]].product).isEqualTo(Currency.EUR)
            a.assertThat(trade.priceCurrency).isEqualTo(Currency.UAH)

            a.assertThat(trade.volume).isEqualTo(bd(1000))
            a.assertThat(trade.price).isEqualTo(Amount.valueOf("36.7 UAH"))
            a.assertThat(trade.tradedAt).isEqualTo(now)

        }.assertAll()
    }

    @Test
    def createStockTrade(): Unit = {
        val trade = StockTrade(
            id = 124L,
            market = TestPredefinedMarkets.NASDAQ.symbol,
            company = TestPredefinedCompanies.APPLE.symbol,
            buySellType = BuySellType.BUY,
            volume = bd("1000"),
            price = Amount.of(bd("95.55"), Currency.USD),
            tradedAt = now,
        )

        SoftAssertions().runTests { a =>

            a.assertThat(trade.id).isEqualTo(124L)
            a.assertThat(trade.market).isNotNull.isEqualTo(TestPredefinedMarkets.NASDAQ.symbol)
            a.assertThat(trade.buySellType).isEqualTo(BuySellType.BUY)
            a.assertThat(trade.product).isEqualTo(TestPredefinedCompanies.APPLE.symbol)
            a.assertThat(trade.company).isEqualTo(TestPredefinedCompanies.APPLE.symbol)
            a.assertThat(trade.volume).isEqualTo(bd(1000))
            a.assertThat(trade.price).isEqualTo(Amount.valueOf("95.55 USD"))
            a.assertThat(trade.tradedAt).isEqualTo(now)

        }.assertAll()
    }

    @Test
    def testTestDateTimeService(): Unit = {
        val dateTimeService = TestDateTimeService(nowSupplier = { now.toInstant })
        assertThat(dateTimeService.now()).isEqualTo(now.toInstant)
    }
}
